pkg/operations/restart.go

```diff
@@ -120,119 +101,55 @@ func (r restartOpsHandler) podApplyCompOps(
 	return !pod.CreationTimestamp.Before(&ops.Status.StartTimestamp)
 }
 
-func (r restartOpsHandler) getComponentOrders(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) ([]opsv1alpha1.ComponentOps, error) {
-	cd := &appsv1.ClusterDefinition{}
-	if opsRes.Cluster.Spec.ClusterDef == "" || opsRes.Cluster.Spec.Topology == "" {
-		return nil, nil
-	}
-	if err := cli.Get(reqCtx.Ctx, client.ObjectKey{Name: opsRes.Cluster.Spec.ClusterDef}, cd); err != nil {
-		return nil, err
-	}
-	// components that require sequential restart
-	var orderedComps []opsv1alpha1.ComponentOps
-	for _, topology := range cd.Spec.Topologies {
-		if topology.Name != opsRes.Cluster.Spec.Topology {
-			continue
+func (r restartOpsHandler) restartComponents(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource, comOpsList []opsv1alpha1.ComponentOps) error {
+	doRestart := func(compSpec *appsv1.ClusterComponentSpec, currCompName, targetCompName string) (bool, error) {
+		if targetCompName != currCompName {
+			return false, nil
 		}
-		if topology.Orders != nil && len(topology.Orders.Update) > 0 {
-			// when using clusterDef and topology, "update orders" includes all components
-			for _, compName := range topology.Orders.Update {
-				// get the ordered components to restart
-				if compOps, ok := r.compOpsHelper.componentOpsSet[compName]; ok {
-					orderedComps = append(orderedComps, compOps.(opsv1alpha1.ComponentOps))
-				}
-			}
+		if r.isRestarted(opsRes, compSpec) {
+			return false, nil
+		}
+		if err := cli.Update(reqCtx.Ctx, opsRes.Cluster); err != nil {
```

Comment:
All modifications to the cluster must be submitted in a single request to avoid breaking the update order of components.